### ISSUE TITLE
Consistently check that selectors are standard before passing to the parser

### ISSUE THREAD

**Issue by k0nserv** [stylelint/stylelint#3130]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Updating to 8.4.0 breaks for us.

> What CSS is needed to reproduce this issue?

```scss
@mixin bpk-locale($locale) {
  html[lang=#{$locale}] &,
  html[lang=#{to_lower_case($locale)}] & {
    @content;
  }
}
```

https://github.com/Skyscanner/backpack/blob/3c53ec780ff922a727923b796be540637ce10e98/packages/bpk-mixins/src/mixins/_utils.scss#L165-L170

> Fails with the following error

```
 166:3  ✖  Cannot parse selector   parseError
```

> Which version of stylelint are you using?

8.4.0
> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

SCSS

> What did you expect to happen?

The linting not to fail

**Comment by jeddy3** [comment]
@k0nserv Thanks for the report and for using the template.

Are you able to paste your config? I believe a particular rule is triggering this error. Help narrowing down which one it could be would be appreciated.

**Comment by k0nserv** [comment]
Hey, thanks for getting back.

This is the [config](https://github.com/Skyscanner/backpack/blob/master/.stylelintrc.json) which extends https://github.com/Skyscanner/stylelint-config-skyscanner/blob/master/index.js

**Comment by k0nserv** [comment]
Is there a way to get verbose output from stylelint? I didn't see any options for it, but I'd be happy to provide more information on the failure if required

**Comment by jeddy3** [comment]
Wow, that's quite the config. It's great to see so many rules and plugins get used.

I did a little digging and there's quite a lot going on here. The following code, i.e. an _unquoted_ attribute selector with non-standard variable interpolation (_containing a function_) and followed by the parent selector, is not parseable postcss-selector-parser:

```scss
a[attribute=#{function($value)}] & {}
```

We normally check that [selectors conform to standard syntax](https://github.com/stylelint/stylelint/blob/master/lib/utils/isStandardSyntaxSelector.js) before passing them along to postcss-selector-parser. This is _not_ happening in numerous rules:

- `selector-attribute-brackets-space-inside`
- `selector-attribute-operator-space-after`
- `selector-attribute-operator-space-before`
- `selector-combinator-space-after`
- `selector-combinator-space-before`
- `selector-descendant-combinator-no-non-space`
- `selector-pseudo-class-parentheses-space-inside`
- `string-quotes`

And in the following plugin:

- `scss/selector-no-redundant-nesting-selector`

The parse error is caught by the built-in rules and displayed as `✖  Cannot parse selector   parseError`. Whereas the plugin doesn't catch the error and is thrown as `Error: Unexpected "(" found.`.

@k0nserv I believe a workaround is to _quote your attribute selectors_ as postcss-selector-parser can parse these. 

Going forward, I think we ought to add `isStandardSyntaxSelector` checks to each of the built-in rules listed above. The authors of `scss/selector-no-redundant-nesting-selector` plugin might want to be less restrictive on non-standard syntax and only ignore selectors that are known to cause postcss-selector-parser problems.

**Comment by k0nserv** [comment]
Thanks for taking the time to look at it. You're right quoting the attributes does help here. Strangely just using `postcss-selector-parser` directly [does not work](https://runkit.com/k0nserv/5a5e0c6cc26f170012f0f473). Making sure that this errors earlier seems like a good idea, the error was a bit confusing as presented to us

**Comment by jeddy3** [comment]
> Thanks for taking the time to look at it.

No worries :)

> Strangely just using postcss-selector-parser directly does not work.

postcss-selector-parser only parsers selector lists and not stylesheets. So, it's going to throw a parse error when given a whole at-rule.

The following is parsed fine:

```js
const parser = require('postcss-selector-parser');
const text = `html[lang="#{to_lower_case($locale)}"] &`;
parser().processSync(text);
```

Whereas it throws without the quotes:

```js
const parser = require('postcss-selector-parser');
const text = `html[lang=#{to_lower_case($locale)}] &`;
parser().processSync(text);
```

> Making sure that this errors earlier seems like a good idea, the error was a bit confusing as presented to us

I think the best thing to do is _ignore_ non-standard syntax selectors in the built-in rules. Once this is done there _shouldn't be any parse errors_ as only standard syntax selectors will be passed on to postcss-selector-parser. Let's wait to see if anyone else has thoughts on this, otherwise I'll label this issue up as help wanted. We've a built-in util for ignoring non-standard syntax selectors, so making these changes to the affected built-in rules should be trivial.

**Comment by jeddy3** [comment]
It looks like there are no objections to my proposal.

@k0nserv If you'd like to use unquoted attribute selectors with non-standard variable interpolation (containing a function) and followed by the parent selector, then please consider improving stylelint by adding the `isStandardSyntaxSelector` check to the rules listed above.

**Comment by brneto** [comment]
I'm facing a problem that I think has to do with this discussion here. I'm trying to use the nested properties syntax in my scss file, but I'm getting the error: `✖ Cannot parse selector parseError`.

I think that's because the `postcss-selector-parser` is considering the code below as a selector and throwing a error:
`
padding: {
  right: 16px;
  bottom: 16px;
  left: 16px;
}
`
Seems that stylelint is not checking the rule `scss/declaration-nested-properties` before passing them along to `postcss-selector-parser` too?
 
@jeddy3, Do you know if would have a way to stylelint lint that syntax correctly?

**Comment by ntwb** [comment]
@brneto As your issue relates to the `scss/declaration-nested-properties` scss rule, could you please check if this has already been reported at https://github.com/kristerkari/stylelint-scss and if not create an issue there please

**Comment by brneto** [comment]
@ntwb I think I've found the problem. It was in the `no-descending-specificity` rule. When it was calling `root.walkRules` the method of PostCSS API it was treating `padding:` as a selector and passing it to parseSelector directly.

I think I fix the problem here and I send a PR, ok?

**Comment by ntwb** [comment]
@brneto Yes please, but also create a separate issue to go with the pull request please

**Comment by brneto** [comment]
Unfortunately, I’ve already created the PR and related to this issue, there’s a way I change that now?!
